radixdlt-core/radixdlt/src/test/java/com/radixdlt/network/p2p/proxy/PeerProxyTest.java

```diff
@@ -236,4 +239,71 @@ public void test_peer_proxy_disabled() throws Exception {
             .getVerifiedProxiesForNode(nodeIdOf(VALIDATOR_NODE));
     assertEquals(0L, verifiedProxies.size());
   }
+
+  @Test
+  public void test_use_proxy_if_direct_connection_is_not_allowed() throws Exception {
+    final var nodeAddressing = Addressing.ofNetwork(Network.LOCALNET).forNodes();
+    final var validatorKey = ECKeyPair.generateNew();
+    final var proxyKey = ECKeyPair.generateNew();
+
+    final var validatorProps = defaultProperties();
+    validatorProps.set(
+        "network.p2p.proxy.authorized_proxies", nodeAddressing.of(proxyKey.getPublicKey()));
+    // validator can only connect to the proxy node
+    validatorProps.set("network.p2p.peer_whitelist", nodeAddressing.of(proxyKey.getPublicKey()));
+
+    final var proxyProps = defaultProperties();
+    proxyProps.set("network.p2p.proxy.enabled", "true");
+    proxyProps.set(
+        "network.p2p.proxy.authorized_proxied_nodes",
+        nodeAddressing.of(validatorKey.getPublicKey()));
+
+    // a network with 3 nodes: a (hidden) validator, a proxy and a default "external" node
+    setupTestRunner(
+        ImmutableList.of(
+            new NodeProps(validatorKey, validatorProps),
+            new NodeProps(proxyKey, proxyProps),
+            new NodeProps(ECKeyPair.generateNew(), defaultProperties())));
+
+    // the validator node has both the proxy and the external node in its address book
+    testNetworkRunner.addressBook(VALIDATOR_NODE).addUncheckedPeers(Set.of(uriOfNode(PROXY_NODE)));
+    testNetworkRunner
+        .addressBook(VALIDATOR_NODE)
+        .addUncheckedPeers(Set.of(uriOfNode(EXTERNAL_NODE)));
+
+    // the proxy node has external node in its address book
+    testNetworkRunner.addressBook(PROXY_NODE).addUncheckedPeers(Set.of(uriOfNode(EXTERNAL_NODE)));
+
+    // the validator sends a message to the external node
+    testNetworkRunner
+        .messageCentral(VALIDATOR_NODE)
+        .send(nodeIdOf(EXTERNAL_NODE), new PeerPingMessage());
+
+    waitForMessagesAndProcessAll();
+
+    // proxy node should forward a single message
+    await()
+        .atMost(Duration.ofSeconds(2))
+        .until(
+            () ->
+                testNetworkRunner.counter(
+                        PROXY_NODE, CounterType.NETWORKING_ROUTING_FORWARDED_MESSAGES)
+                    == 1L);
+
+    // validator should only be connected to the proxy node
+    final var connectedPeers =
+        testNetworkRunner.peerManager(VALIDATOR_NODE).activeChannels().stream()
+            .map(PeerChannel::getRemoteNodeId)
+            .collect(Collectors.toSet());
+    assertTrue(connectedPeers.contains(nodeIdOf(PROXY_NODE)));
+    assertFalse(connectedPeers.contains(nodeIdOf(EXTERNAL_NODE)));
```

Comment:
Is it worth testing that the External Node can send a message to the Validator via the proxy?

In fact, does the External Node know about the Validator at all if it's not in its address book? Do we actually just need to hide the IP when sharing the address book contents?

Reply:
The external node won't have an entry for the validator node in its address book but it will have an entry for the proxy node and a corresponding proxy certificate (issued and signed by the validator node).

Reply:
> Is it worth testing that the External Node can send a message to the Validator via the proxy?

I've extended the test